### PR TITLE
#2639 Temperature sync scanning

### LINKED ISSUE #2639: Temperature syncing logic: scanning for sensors
## Is your feature request related to a problem? Please describe.

Related issue #2600 

To do all use-cases of the temperature syncing in a single issue is a bit much. This issue will be the code for the actions to scan for sensors

## Describe the solution you'd like

- A `TemperatureReducer` connected to redux
- A `startScan` action
- A `completeScan` action
- A `scanError` action
- A `saveSensors` action
- A `scanForSensors` action (composition of the above actions)

## Implementation

- As above

## Describe alternatives you've considered

N/A

## Additional context

N/A

### PR DESCRIPTION
**Pull request by joshxg**
### NOTE: Branched from #2639 

Fixes #2639 

## Change summary

- Adds actions for scanning for sensors and saving new/updating old
- Adds an initial `TemperatureSyncReducer`

## Testing

N/A

### Related areas to think about

There is possibilities for some exceptions to be thrown - going to move forward and see if I can find them all (there are a few going to come up with the app being put into the background etc etc when the scheduling is being run n stuff)

### CODE REVIEW

File: src/actions/TemperatureSyncActions.js [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+  if (!isArray) dispatch(scanError());
+  if (isArray && !sensorAdvertisements.length) dispatch(scanError());
+
+  if (isArray && sensorAdvertisements.length) {
+    sensorAdvertisements.forEach(({ macAddress, batteryLevel }) => {
+      UIDatabase.write(() => {
+        UIDatabase.update('Sensor', {
+          id: generateUUID(),
+          ...UIDatabase.get('Sensor', 'macAddress', macAddress),
```
Comment (joshxg) [inline src/actions/TemperatureSyncActions.js:32]: the second and third parameters of this are backwards, oops! - fixed in #2650

File: src/actions/TemperatureSyncActions.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  SCAN_START: 'TemperatureSync/scanStart',
+  SCAN_COMPLETE: 'TemperatureSync/scanComplete',
+  SCAN_ERROR: 'TemperatureSync/scanError',
+};
+
+const scanStart = () => ({ type: TEMPERATURE_SYNC_ACTIONS.SCAN_START });
+const scanComplete = () => ({ type: TEMPERATURE_SYNC_ACTIONS.SCAN_COMPLETE });
+const scanError = () => ({ type: TEMPERATURE_SYNC_ACTIONS.SCAN_ERROR });
+
+const updateSensors = sensorAdvertisements => dispatch => {
```
Comment (joshxg) [inline src/actions/TemperatureSyncActions.js:21]: this is working well. A good advantage to this is we don't need to throw up the spinner while it's scanning. In the page where we display and hook up sensors to locations, we can hook into the `scanComplete()` action to refresh that page with the new sensors

File: src/reducers/TemperatureSyncReducer.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  SCANNING: 'SCANNING',
+  SCAN_ERROR: 'SCAN_ERROR',
+};
+
+const initialState = () => ({
+  progress: 0,
+  total: 0,
+  syncState: null,
+  isSyncing: false,
+});
```
Comment (joshxg) [inline src/reducers/TemperatureSyncReducer.js:18]: really basic state at the moment. The progress/total should just be for downloading logs, and the `syncState` can possibly help the UI determine what text to display, but will see if that's enough. Trying to have the reducer state reasonably generic where you could chuck it into another app quite easily

File: src/actions/TemperatureSyncActions.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+  const isArray = Array.isArray(sensorAdvertisements);
+
+  if (!isArray) dispatch(scanError());
```
Comment (mark-prins) [inline src/actions/TemperatureSyncActions.js:24]: I like early returns. I would do this:

```suggestion
  if (!isArray) {
    dispatch(scanError());
    return;
 }
```

File: src/actions/TemperatureSyncActions.js [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+  if (isArray && !sensorAdvertisements.length) dispatch(scanError());
```
Comment (mark-prins) [inline src/actions/TemperatureSyncActions.js:25]: and then this:
```suggestion
  if (isArray && !sensorAdvertisements.length) {
    dispatch(scanError());
    return;
  }
```

File: src/actions/TemperatureSyncActions.js [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+  if (isArray && sensorAdvertisements.length) {
```
Comment (mark-prins) [inline src/actions/TemperatureSyncActions.js:27]: and then remove this `if`

**Comment by joshxg** [comment]
yeah definitely - address in here: https://github.com/openmsupply/mobile/issues/2665 :)

**Pull request merged**